meshmode/mesh/refinement/tesselate.py

```diff
@@ -23,105 +24,283 @@
-
-    node_dict = {
-          ituple: idx
-          for idx, ituple in enumerate(node_tuples)}
-
-    def try_add_tet(current, d1, d2, d3, d4):
-        try:
-            result.append((
-                node_dict[add_tuples(current, d1)],
-                node_dict[add_tuples(current, d2)],
-                node_dict[add_tuples(current, d3)],
-                node_dict[add_tuples(current, d4)],
-                ))
-        except KeyError:
-            pass
-
-    result = []
-
-    if len(result) > 0:
-        return [node_tuples, result]
-    for current in node_tuples:
-        # this is a tesselation of a cube into six tets.
-        # subtets that fall outside of the master tet are simply not added.
-
-        # positively oriented
-        try_add_tet(current, (0, 0, 0), (1, 0, 0), (0, 1, 0), (0, 0, 1))
-        try_add_tet(current, (1, 0, 1), (1, 0, 0), (0, 0, 1), (0, 1, 0))
-        try_add_tet(current, (1, 0, 1), (0, 1, 1), (0, 1, 0), (0, 0, 1))
-
-        try_add_tet(current, (1, 0, 0), (0, 1, 0), (1, 0, 1), (1, 1, 0))
-        try_add_tet(current, (0, 1, 1), (0, 1, 0), (1, 1, 0), (1, 0, 1))
-        try_add_tet(current, (0, 1, 1), (1, 1, 1), (1, 0, 1), (1, 1, 0))
-
-    return node_tuples, result
-
-
-def tesselate_simplex_bisection(dim):
-    if dim == 1:
-        return tesselateseg()
-    elif dim == 2:
-        return tesselatetri()
-    elif dim == 3:
-        return tesselatetet()
+    return tuple(midpoint(ai, bi) for ai, bi in zip(a, b))
+
+
+# {{{ shape-dependent tesselation helpers
+
+@singledispatch
+def get_child_basis_vertex_indices(shape: mp.Shape, child):
+    assert len(child) == shape.nvertices
+    return child[:shape.dim + 1]
+
+
+@get_child_basis_vertex_indices.register(mp.Hypercube)
+def _(shape: mp.Hypercube, child):
+    assert len(child) == shape.nvertices
+
+    # * for a cube, we reorder nodes such that the vectors
+    #
+    #       (0, 1), (0, 2) and (0, 4)
+    #
+    #   form an orthogonal basis, since (0, 3) is linearly dependent on 1 and 2.
+    #
+    # * lines and squares don't require any reordering
+
+    return [child[i] for i in [0, 1, 2, 4][:shape.dim + 1]]
+
+# }}}
+
+
+# {{{ resampling
+
+def get_ref_midpoints(shape, ref_vertices):
+    r"""The reference element is considered to be, e.g. for a 2 simplex::
+
+        F
+        | \
+        |   \
+        D----E
+        |   /| \
+        | /  |   \
+        A----B----C
+
+    where the midpoints are ``(B, E, D)``. The same applies to other shapes
+    and higher dimensions.
+
+    :arg ref_vertices: a :class:`list` of node index :class:`tuple`\ s
+        on :math:`[0, 2]^d`.
+    """
+
+    from pytools import add_tuples
+    space = mp.space_for_shape(shape, 1)
+    orig_vertices = [
+            add_tuples(vt, vt) for vt in mp.node_tuples_for_space(space)
+            ]
+    return [rv for rv in ref_vertices if rv not in orig_vertices]
+
+
+def map_unit_nodes_to_children(shape, unit_nodes, tesselation):
```

Comment:
```suggestion
def map_unit_nodes_to_children(shape, unit_nodes, tess_info):
```
?
(Here and elsewhere)

Reply:
Done in 4537029.